website/docs/generics.md

```diff
@@ -0,0 +1,1102 @@
+---
+id: generics
+title: Generic Classes and Methods
+sidebar_label: Generics
+---
+
+Sorbet has syntax for creating generic methods, classes, and interfaces.
+
+## Implementation status
+
+Generics are the most poorly supported feature within Sorbet, by far. Their
+implementation has many known, complicated-to-fix bugs which substantially limit
+their utility.
+
+It is hard to overstate how buggy Sorbet's support for generics is.
+
+We track known bugs in Sorbet's support for generics in the [Generics milestone]
+on Sorbet's issue tracker. Feel free to peruse the existing bugs, as well as
+report new bugs.
+
+Unfortunately, this abundance of bugs makes implementing sophisticated, generic
+abstractions in Sorbet exceedingly brittle.
```

Comment:
I understand the desire to emphasize the shortcomings of Sorbet's generics support.  (I think internal discussions had suggested separating "generic classes" from "generic methods" as the former are relatively more stable, but maybe you made the -- reasonable! -- executive decision to combine them, as it's hard to write generic classes without also extensively using generic methods.)

But the practical reality is that people use Sorbet's generics support every single day via the standard library, and Stripe's codebase has thousands of instances of `type_member`/`type_template` even prior to docs on generics being written.  People do run into issues with the generics support (cf. the generics milestone, etc.), yes, and there's no guarantee that all of the instances of `type_member` et al have been rigorously subjected to the kind of tests the documentation below suggests.  And yet here we are, writing the longest piece of documentation (by lines) that exists on Sorbet's website.

I don't know what the right way to modify this introduction is, or even if it should be modified.  But I do think that acknowledging the usefulness of generics while being forceful about their shortcomings would leave fewer people scratching their heads about why this documentation exists.

(I also think "It is hard to overstate how buggy Sorbet's support for generics is." is unnecessary.)

Reply:
The docs are long not because it is so widely used, but because it requires so much knowledge to use correctly.

The prevalence of usage in Stripe's codebase is not very diverse. Almost all of them blame to one or two narrowly defined abstractions (DataViews). Some caveats about that:

- They actually do suffer from nearly all the problems mentioned in this doc. Nearly 100% of the implementation of DataView is untyped to work around the bugs.
- Child classes almost never touch those annotations—the annotations are kept in sync by automated scripts.
- The core abstraction was written very carefully in coordination with members of the Sorbet team, and has not been changed much since. Nearly no one else using generics will have as intimate an understanding of the limitations of Sorbet's generics when building new abstractions.

I have removed the "hard to overstate" line, but I 100% still believe it is accurate.

The purpose of this part of the doc is intentionally to scare people into testing that their generic abstractions _actually work_, rather than inflicting pain on unsuspecting consumers.

If you have further suggestions, please lmk.

Reply:
Some reasons why I wrote one page:

- it's easier to ⌘F a single page, and generic methods vs generic classes is not something i expect people to have a solid understanding of
- i only wanted to write the disclaimer once, not across two pages
- you're right that generic classes do frequently also want generic methods (though as this doc does demonstrate, it is possible to have examples that use one or the other entirely, not both)